pkg/services/searchV2/readme.md

```diff
@@ -0,0 +1,98 @@
+Here is a proposal about a new package layout which:
+
+* will make different types of indexes work separately, each consuming `entity_events` individually in its own control goroutine, so less contention
+* will allow doing re-indexing with different periods for different types of indexes
+* will allow doing isolated backups, will simplify extending in custom way
+
+Entities:
+
+* `StandardSearchService` - wrapper for all known indexes. For now only has reference to `dashboardSearch`.
+* `dashboardSearch` - manages dashboard indexes for all organizations, every organization has its own `dashboardIndex`.
+* every `dashboardIndex` has its own `*bluge.Writer`
+
+```
+type StandardSearchService struct {
+  dashboardSearch *dashboardSearch
+}
+
+type dashboardSearch struct {
+  indexes map[int64]*dashboardIndex
+}
+
+type dashboardIndex struct {
+  writer *bluge.Writer
+}
+
+func (*dashboardSearch) run(ctx) error {
+  // Create initial indexes for known orgs.
+  // Start applying events, do full re-indexing, do backups etc.
+}
+
+func (*dashboardSearch) backup(ctx, currentEventID) error {
+  // write meta.json
+  // for each org do the backup using *bluge.Writer.
+  // done.
+}
+```
+
+Backup directory is configurable per index type. Inside backup directory structure may look like this:
+
+```
+dashboard
+   ├-- meta.json
+   ├── org1
+   │   ├── 0000000004a7.seg
+   │   ├── 0000000004a8.seg
+   │   ├── 0000000004a9.seg
+   │   └── 000000000893.snp
+   └── org2
+```
+
+`meta.json` contains:
+
+```
+{
+  "eventId": 12
+}
+```
+
+* `eventId` - the last event ID applied to the index in the backup. Since we make backups periodically we may need to apply some missing updates from `entity_event` table to catch up the state.
+
+As we don't have org id separation in `entity_event` table we manage indexes for all organizations in one goroutine. By different types of indexes are a separate consumers of `entity_event` table - so different types of indexes do not depend on each other at all.
+
+`IndexManager` is responsible controlling index lifecycle (actually this is our current `run` method), it accepts interface like:
+
+```
+type Index interface {
+  Init(ctx, orgIDs) error // Index initially or load from existing backup.
+  FullReIndex(ctx) error
+  HasOrgIndex(ctx, orgID) bool
+  ReIndexForOrg(ctx, orgID) error
+  ApplyUpdates(ctx, []EntityEvent) newEventID
+  DoBackup(ctx, currentEventID) error
+}
+```
+
+Index implementation may re-use `BackupManager`:
+
+```
+type BackupData struct {
+  Indexes map[int64]*bluge.Writer
+  EventID int64
+}
+
+type BackupManager interface {
+  Save(BackupData) error
+  Load() (BackupData, bool, error)
+}
+```
+
+We can do backups after full-reindexing with the event id seen before re-indexing started.
+
+Notes:
+
+* At this moment we still need to re-index periodically since not all changes come to `entity_event` table.
```

Comment:
Will backups ever be useful before we 
- publish change events atomically, and
- assure we publish events on _all_ changes
?

Reply:
Also thought on this, think the answer is yes - will be useful anyway, because we only save them after full reIndexing for now, with eventID set to the one before re-indexing. So it should be consistent.

Reply:
Lets say the following happens:

1. full reindex
2. backup
3. event #1
4. event #2
5. event #3
6. node failure
7. restart -> restore backup or full reindex?

In this case, wouldn't we still need to do full reindex rather than restore? Otherwise we would have missed some updates

Reply:
We will re-apply 3, 4, 5 after node restart. But in the worst case scenario we can still miss some updates not presented in entity_events table – in this case we will live with non-actual data until next re-index attempt, but this could happen anyway without node restart involved.

Reply:
right, I missed that backup contains the last received event id:

1. full reindex
2. backup - last event id `123`
3. event no.124
4. event no.125
5. event no.126
6. node failure
7. restart -> restore backup and retrieve all events after 123

Reply:
Added